library/ccm.c

```diff
@@ -291,60 +341,104 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
     len_left = length;
     src = input;
     dst = output;
-
+    if( len_left == 0 )
+    {
+        fc3 = 0;
+        fc4 = 0;
+        fc5 = 0;
+    }
     while( len_left > 0 )
     {
         size_t use_len = len_left > 16 ? 16 : len_left;
 
         if( mode == CCM_ENCRYPT )
         {
-            mbedtls_platform_memset( b, 0, 16 );
-            mbedtls_platform_memcpy( b, src, use_len );
-            UPDATE_CBC_MAC;
+            mbedtls_platform_random_delay();
+            if( mode_dup == CCM_ENCRYPT )
+            {
+                mbedtls_platform_memset( b, 0, 16 );
+                mbedtls_platform_memcpy( b, src, use_len );
+                UPDATE_CBC_MAC;
+                CHECK_FLOW_CTRL( 16, fc3, flow_ctrl_local );
+                fc5 = 0;
+            }
+            else
+            {
+                return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+            }
         }
 
         CTR_CRYPT( dst, src, use_len );
+        CHECK_FLOW_CTRL( use_len, fc4, flow_ctrl_local );
 
         if( mode == CCM_DECRYPT )
         {
-            mbedtls_platform_memset( b, 0, 16 );
-            mbedtls_platform_memcpy( b, dst, use_len );
-            UPDATE_CBC_MAC;
+            mbedtls_platform_random_delay();
+            if( mode_dup == CCM_DECRYPT )
+            {
+                mbedtls_platform_memset( b, 0, 16 );
+                mbedtls_platform_memcpy( b, dst, use_len );
+                UPDATE_CBC_MAC;
+                CHECK_FLOW_CTRL( 16, fc5, flow_ctrl_local );
+                fc3 = 0;
+            }
+            else
+            {
+                return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+            }
         }
 
         dst += use_len;
         src += use_len;
         len_left -= use_len;
-
         /*
          * Increment counter.
          * No need to check for overflow thanks to the length check above.
          */
         for( i = 0; i < q; i++ )
             if( ++ctr[15-i] != 0 )
                 break;
+        for( i = 0; i < q; i++ )
+            if( ++ctr_dup[15-i] != 0 )
+                break;
+
+        if( mbedtls_platform_memcmp( ctr_dup, ctr, 16 ) != 0 )
+            return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
     }
 
+    if( len_left > 0 )
+    {
+        return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+    }
     /*
      * Authentication: reset counter and crypt/mask internal tag
      */
+    flow_ctrl_local = 0;
     for( i = 0; i < q; i++ )
+    {
+        flow_ctrl_local++;
```

Comment:
@AndrzejKurek , can we use secured memset instead of for and flow counter ?
If we want to use for, we need 1 loop per ctr and another for ctr_dup

Reply:
Added.